storage/database/src/backend/heed/env.rs

```diff
@@ -282,10 +282,32 @@ where
         tx_rw: &RefCell<heed::RwTxn<'env>>,
     ) -> Result<impl DatabaseRw<T>, RuntimeError> {
         // Open up a read/write database using our table's const metadata.
-        Ok(HeedTableRw {
-            db: self.create_database(&mut tx_rw.borrow_mut(), Some(T::NAME))?,
-            tx_rw,
-        })
+        let db = {
+            let mut tx_rw = tx_rw.borrow_mut();
+            let mut db = self.database_options();
+
+            // Set the key comparison behavior.
+            match <T::Key>::KEY_COMPARE {
+                // Already the default for LMDB, setting
+                // a custom comparison function is slower.
+                KeyCompare::Lexicographic => (),
+
+                // Instead of setting a custom [`heed::Comparator`],
+                // use this LMDB flag; it is ~10% faster.
+                KeyCompare::Number => {
+                    db.flags(DatabaseFlags::INTEGER_KEY);
+                }
+
+                // Use a custom comparison function if specified.
+                KeyCompare::Custom(_) => {
+                    db.key_comparator::<T::Key>();
+                }
+            }
```

Comment:
@Boog900 There is a problem with this.

Since the comparator function is baked into heed's database _type_, setting a different one means we're returning a different type - we can't do this since `-> impl DatabaseRw<T>` must be _one_ type.

I.e. in the first 2 branches we're returning:
```rust
heed::Database<K, V, heed::DefaultComparator>
```
but in the last, we're returning:
```rust
//            implements heed::Comparator
//                   v
heed::Database<K, V, StorableHeed<Key>>
```

Workarounds I've thought about so far:

Workaround: always call `key_comparator`, the `heed::Comparator` impl forwards to the correct fn at compile time
Con: ~10% slower DB writes, even the default. I believe it's because LMDB has to call our (un-inline-able?) function instead of using its built-in one

Workaround: Only support `u32, u64, usize` for value comparisons
Con: No custom comparison for anything other than those

Workaround (best IMO): Document an invariant that the new `create_db` _must_ be called to "set" the flags / comparison fns. After that, `open_db_r{o,w}` will be like before this PR - open with no flags or comparison function, assume they were set with `create_db`.
Con: Users must follow the invariant for correct behavior.

The last one works since `create_db` doesn't return the DB, so no type worries.

Reply:
Ah ok, yeah I like the the last option best as well

> Con: ~10% slower DB writes,

LMDB still needs to call the comparison function for reads, so reads should be slower as well, probably more so as the comparison function takes up a bigger percent of the total work for reads compared to writes.